src/parser/core/modules/Invulnerability.tsx

```diff
@@ -549,7 +549,16 @@ export class Invulnerability extends Module {
 		end = this.parser.currentTimestamp,
 	) => this.getInvulnUptime(targetId, start, end, 'invulnerable')
 
-	getInvulns(
+	// Public interface for invuln windows that ensures consumers can't look ahead to the future
+	getInvulns = (...args: Parameters<typeof Invulnerability.prototype.getInvulnsInternal>) =>
+		this.getInvulnsInternal(...args)
+			.filter(window => window.start <= this.parser.currentTimestamp)
+			.map(window => ({
+				...window,
+				end: Math.min(window.end, this.parser.currentTimestamp),
+			}))
```

Comment:
we need the full data set for normalisation purposes on this current impl, so I've kept the logic as-is, and added this wrapper which exposes only what's technically possible during the main run.